File: Python-codes/simple_linear_regression.py
from sklearn.datasets import load_diabetes # data set
from sklearn.linear_model import LinearRegression # model fitting
from sklearn.metrics import mean_squared_error, r2_score # mse and r^2
import numpy as np
import matplotlib.pyplot as plt # data visualization

# data set
diabetes = load_diabetes()

# data manipulation
X_ = diabetes.data
y = diabetes.target
X = X_[:, np.newaxis, 2] # only 1 feature (it is simple linear regression)

# shuffling data
n_samples = X.shape[0]
train_size = int(n_samples * 0.8)
indices = np.random.permutation(n_samples)
train_idx = indices[:train_size]
test_idx = indices[train_size:]

# data set splitting
X_train = X[train_idx]
y_train = y[train_idx]
X_test = X[test_idx]
y_test = y[test_idx]

# model fitting
linear_reg_model = LinearRegression()
linear_reg_model.fit(X_train, y_train)

# y_pred
y_pred = linear_reg_model.predict(X_test)

# mse and r^2
mse = mean_squared_error(y_test, y_pred)
print("mse: ", mse)
r2 = r2_score(y_test, y_pred)
print("r^2: ", r2)

# data visualization
plt.scatter(X_test, y_test, color = "gray")
plt.plot(X_test, y_pred, color = "red")

File: R-codes/simple_linear_regression.R
# data set
library(lars)
data(diabetes)                # from lars package
X_full <- diabetes$x          # 10 standardized predictors
y <- diabetes$y               # target (quantitative)

# select a single feature (for simple linear regression)
X <- X_full[, 3, drop = FALSE]
colnames(X) <- "feature"

# shuffling & splitting (80/20)
set.seed(42)   # ensures reproducibility of the random split

n_samples <- nrow(X)
train_size <- floor(0.8 * n_samples)
indices <- sample(n_samples)
train_idx <- indices[1:train_size]
test_idx <- indices[(train_size + 1):n_samples]

X_train <- X[train_idx, , drop = FALSE]
y_train <- y[train_idx]
X_test <- X[test_idx, , drop = FALSE]
y_test <- y[test_idx]

# model fitting
df_train <- data.frame(y = y_train, feature = X_train[, 1])
lm_model <- lm(y ~ feature, data = df_train)

# prediction
y_pred <- predict(lm_model, newdata = data.frame(feature = X_test[, 1]))

# metrics: MSE and R^2 on test set
mse <- mean((y_test - y_pred)^2)
ss_res <- sum((y_test - y_pred)^2)
ss_tot <- sum((y_test - mean(y_test))^2)
r2 <- 1 - ss_res / ss_tot

cat("mse: ", mse, "\n")
cat("r^2: ", r2,  "\n")

# baseR: simple data visualization
plot(X_test[, 1], y_test,
     col = "gray", pch = 16,
     xlab = "Feature (Column 3)", ylab = "Target",
     main = "Simple Linear Regression (R)")
abline(lm_model, col = "red", lwd = 2)
